### PR TITLE
7536 use head for snap sync

### LINKED ISSUE #7536: Use head block instead of safe block for snap sync
Change snap sync to use head block instead of safe block for the snap sync.

Using the head allows for larger download windows between two pivot blocks. With larger windows, we will have fewer pivot blocks and therefore more time for the heal. This could allow us to finish faster. If we download the storage of an account and then change the pivot block, we need to request the account again to get the new storage root of the account in the block, as this can cause a lot of storage requests to fail.

Acceptance Criteria
* Verify that this has the improvement on healing and syncing that we are expecting

### PR DESCRIPTION
**Pull request by kingnhcomcast**
## Fixed Issue(s)

fixes #7536

Default is sync to head, added config option 'Xsnapsync-to-head-enabled' which can be set to false to revert to older sync to safe block.

### Thanks for sending a pull request! Have you done the following?

- [ ] Checked out our [contribution guidelines](https://github.com/hyperledger/besu/blob/main/CONTRIBUTING.md)?
- [ ] Considered documentation and added the `doc-change-required` label to this PR [if updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).
- [ ] Considered the changelog and included an [update if required](https://wiki.hyperledger.org/display/BESU/Changelog).
- [ ] For database changes (e.g. KeyValueSegmentIdentifier) considered compatibility and performed forwards and backwards compatibility tests

### Locally, you can run these tests to catch failures early:

- [ ] unit tests: `./gradlew build`
- [ ] acceptance tests: `./gradlew acceptanceTest`
- [ ] integration tests: `./gradlew integrationTest`
- [ ] reference tests: `./gradlew ethereum:referenceTests:referenceTests`

### CODE REVIEW

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/PivotSelectorFromBlock.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+                            }))
+            .orElse(0L),
+        localChainHeight);
+  }
+
+  @Override
+  public Optional<FastSyncState> selectNewPivotBlock() {
+    final Optional<ForkchoiceEvent> maybeForkchoice = forkchoiceStateSupplier.get();
+    if (maybeForkchoice.isPresent() && maybeForkchoice.get().hasValidSafeBlockHash()) {
+      // return Optional.of(selectBlockAsPivot(maybeForkchoice.get().getSafeBlockHash()));
```
Comment (garyschulte) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/PivotSelectorFromBlock.java:128]: nit: remove old commented code

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/PivotSelectorFromBlock.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+                              return null;
+                            }))
+            .orElse(0L),
+        localChainHeight);
+  }
+
+  @Override
+  public Optional<FastSyncState> selectNewPivotBlock() {
+    final Optional<ForkchoiceEvent> maybeForkchoice = forkchoiceStateSupplier.get();
+    if (maybeForkchoice.isPresent() && maybeForkchoice.get().hasValidSafeBlockHash()) {
```
Comment (garyschulte) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/PivotSelectorFromBlock.java:127]: hasValidSafeBlockHash() isn't necessary for the head block implementation.  I would suggest making getPivotHash return Optional<Hash>, and you can do the safe block check there and return Optional.empty() if it doesn't exist.

this is a very corner case, where we would be trying to sync a network very near genesis, before a block has been marked 'safe'.  The prior implementation of safe-only would have the same issue.  So this isn't a huge deal, but fixing it now might save a headache in the future for sync tests or something like that where we would be syncing a brand new chain.

**Review by garyschulte — CHANGES_REQUESTED** [review]
LGTM, good tests.  One nit and one suggestion

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/PivotSelectorFromBlock.java (hunk 2 above)
Comment (kingnhcomcast) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/PivotSelectorFromBlock.java]: code removed

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/PivotSelectorFromBlock.java (hunk 1 above)
Comment (kingnhcomcast) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/PivotSelectorFromBlock.java]: code removed

**Review by garyschulte — APPROVED** [review]
🚢

**Pull request merged**